app/streak/FilterBuilderInput.tsx

```diff
@@ -0,0 +1,351 @@
+          value={inputText}
+          onChange={handleInputChange}
+          onFocus={() => setIsFocused(true)}
+          // onBlur={() => setTimeout(() => setIsFocused(false), 200)} // Delay to allow click on options
+          onKeyDown={handleKeyDown}
+          placeholder="Type to search..."
+          className="w-full border-none p-2 text-white focus:outline-none"
+        />
+
+        {/* Floating Options Div */}
+        {isFocused && filteredOptions.length > 0 && (
+          <div
+            ref={optionsRef}
+            className="absolute z-10 mt-1 max-h-48 w-full overflow-y-auto rounded-md border border-gray-300 bg-white shadow-lg dark:bg-gray-900"
+          >
+            <ul>
+              {currFilter === null &&
+                filteredFilterDefOptions.map((filterDef, index) => (
+                  <li
+                    aria-label={`Option: ${filterDef.filterKey}`}
+                    key={filterDef.filterKey}
+                    onClick={() => handleFilterSelect(filterDef)}
+                    className={`cursor-pointer p-2 hover:bg-gray-200 dark:hover:bg-gray-700 ${
+                      index === activeOptionIndex
+                        ? 'bg-gray-200 dark:bg-gray-700'
+                        : ''
+                    }`}
+                  >
+                    {filterDef.filterKey}
+                  </li>
+                ))}
+              {currFilter !== null &&
+                currRelOperator === null &&
+                [...new Set(filteredRelOpOptions)].map((operator, index) => (
+                  <li
+                    aria-label={`Option: ${operator}`}
+                    key={operator}
+                    onClick={() => handleRelationalOperatorSelect(operator)}
+                    className={`cursor-pointer p-2 hover:bg-gray-200 dark:hover:bg-gray-700 ${
+                      index === activeOptionIndex
+                        ? 'bg-gray-200 dark:bg-gray-700'
+                        : ''
+                    }`}
+                  >
+                    {operator}
+                  </li>
+                ))}
+              {currFilter !== null &&
+                currRelOperator !== null &&
+                loadedOptions?.length &&
+                currFilter.optionKey &&
+                loadedOptions.map((option, index) => (
+                  <li
+                    aria-label={`Option: ${option[currFilter.optionKey]}`}
+                    key={option[currFilter.optionKey]}
+                    onClick={() =>
+                      handleValueSelection([option[currFilter.optionKey]])
+                    }
+                    className={`cursor-pointer p-2 hover:bg-gray-200 dark:hover:bg-gray-700 ${
+                      index === activeOptionIndex
+                        ? 'bg-gray-200 dark:bg-gray-700'
+                        : ''
+                    }`}
+                  >
+                    {option[currFilter.optionKey]}
+                  </li>
```

Comment:
I should combine all of these. Lots of duplication that could be handled by a bit of logic to determine which handler should be used and which text...